Lib/fontmake/instantiator.py

```diff
@@ -0,0 +1,545 @@
+        raise ValueError(
+            f"Cannot swap glyphs '{name_old}' and '{name_new}', as either or both are "
+            "missing."
+        )
+
+    # 1. Swap outlines and glyph width. Ignore lib content and other properties.
+    glyph_swap = ufoLib2.objects.Glyph(name="temporary_swap_glyph")
+    glyph_old = font[name_old]
+    glyph_new = font[name_new]
+
+    p = glyph_swap.getPointPen()
+    glyph_old.drawPoints(p)
+    glyph_swap.width = glyph_old.width
+
+    glyph_old.clear()
+    p = glyph_old.getPointPen()
+    glyph_new.drawPoints(p)
+    glyph_old.width = glyph_new.width
+
+    glyph_new.clear()
+    p = glyph_new.getPointPen()
+    glyph_swap.drawPoints(p)
+    glyph_new.width = glyph_swap.width
+
+    # 2. Remap components.
+    for g in font:
+        for c in g.components:
+            if c.baseGlyph == name_old:
+                c.baseGlyph = name_new
+            elif c.baseGlyph == name_new:
+                c.baseGlyph = name_old
+
+    # 3. Swap literal names in kerning.
+    kerning_new = {}
+    for first, second in font.kerning.keys():
+        value = font.kerning[(first, second)]
+        if first == name_old:
+            first = name_new
+        elif first == name_new:
+            first = name_old
+        if second == name_old:
+            second = name_new
+        elif second == name_new:
+            second = name_old
+        kerning_new[(first, second)] = value
+    font.kerning = kerning_new
+
+    # 4. Swap names in groups.
+    for group_name, group_members in font.groups.items():
+        group_members_new = []
+        for name in group_members:
+            if name == name_old:
+                group_members_new.append(name_new)
+            elif name == name_new:
+                group_members_new.append(name_old)
+            else:
+                group_members_new.append(name)
+        font.groups[group_name] = group_members_new
+
+
+@attr.s(auto_attribs=True, frozen=True, slots=True)
+class Variator:
+    """A middle-man class that ingests a mapping of normalized locations to
+    masters plus axis definitions and uses varLib to spit out interpolated
+    instances at specified normalized locations.
+
+    fontMath objects stand in for the actual master objects from the
+    UFO. Upon generating an instance, these objects have to be extracted
+    into an actual UFO object.
+    """
+
+    masters: List[FontMathObject]
+    location_to_master: Dict[Tuple[Tuple[str, float], ...], FontMathObject]
+    model: varLib.models.VariationModel
+
+    @classmethod
+    def from_masters(
+        cls, items: List[Tuple[Location, FontMathObject]], axis_order: List[str]
+    ):
+        masters = []
+        master_locations = []
+        location_to_master = {}
+        for normalized_location, master in items:
+            master_locations.append(normalized_location)
+            masters.append(master)
+            location_to_master[tuple(normalized_location.items())] = master
```

Comment:
should you maybe sort the `normalized_location.items()` before casting that to a tuple? in case the order happens to be different?

Reply:
Mh, yes. Another thing is maybe constructing a location tuple in generate_instance so we don't have to regenerate it every time we call a Variator :thinking:

Reply:
Urgh, wait, VariationModel expects dict location parameters. Meh. I guess I can attach a `lru_cache` if this turns out to be a memory/performance waster.